[WIP] SourceMorph (GSoC)

TommyClausner opened the pull request:
This PR is about updating the SourceMorph class to morph SourceEstimate + tutorials and examples.

To do:
- [ ] hope for checks
- [ ] convince people to merge :)

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -122,57 +117,14 @@ def test_volume_stc():
-                            [stc.vertices, stc.vertices],
-                            tmin=stc.tmin, tstep=stc.tstep)
-    img = stc.as_volume(src, dest='mri', mri_resolution=False)
-    assert (img.shape == src[0]['shape'] + (len(stc.times),))
```
@TommyClausner did you move this test?

if you modify the existing tests it makes it hard to know if you broke something.
I therefore increases the risk unnecessarily.

TommyClausner replied:
but but but wait... I mean, yes I moved the tests, but don't you think this makes sense? I mean I didn't modify them except for replacing assert_equal with assert

agramfort replied:
yes it makes sense except it makes reviewing more difficult. So we can do this in 2 steps if it helps reviewing process.

agramfort replied:
honestly I think it would be safer to put them back to where they were. You can use the -k option of pytest to decide to run a specific test. I want to avoid us overlooking an API breakage.

thanks

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -283,8 +229,8 @@ def test_io_stc():
-    stc.save(op.join(tempdir, "tmp.stc"))
-    stc2 = read_source_estimate(op.join(tempdir, "tmp.stc"))
+    stc.save(os.path.join(tempdir, "tmp.stc"))
+    stc2 = read_source_estimate(os.path.join(tempdir, "tmp.stc"))
```
please avoid cosmetic changes. The diff is already super huge. Can I ask you to undo this?

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -543,125 +490,6 @@ def test_extract_label_time_course():
-    stc_vec_to1 = stc_vec.morph(subject_to, grade=3, smooth=12,
-                                buffer_size=1000, subjects_dir=subjects_dir)
-    stc_vec_to2 = stc_vec.morph_precomputed(subject_to, vertices_to, morph_mat)
-    assert_array_almost_equal(stc_vec_to1.data, stc_vec_to2.data)
```
same remark. I would put it back.

agramfort replied:
unless you confirm that you just copy pasted into test_morph.py

TommyClausner replied:
I mean I copy pasted and removed the buffer size, set the asserts to the new default...

agramfort commented on mne/source_estimate.py:
```diff
@@ -2454,260 +2249,52 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
+        source_morph.data.update({str(hemi): stc_from.vertices[hemi]})
+    source_morph.data.update({'hemis': hemis})
+    source_morph.data.update(_compute_morph_data(source_morph,
+                                                 verbose=verbose))
```
this is a code smell. You create an object but then you need to update its internal state without using a method of the object. Can we avoid this?

agramfort commented on mne/source_estimate.py:
```diff
@@ -2725,20 +2312,14 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
+    source_morph.data['vertno'] = vertices_to
+    source_morph.data['morph_mat'] = morph_mat
+
+    stc_to = _apply_morph_data(source_morph, stc_from)
```
same remark. Could we make SourceMorph class more self sufficient?

agramfort replied:
why not writing

`source_morph(stc_from)`

?

that seems the most natural way for writing it.

regarding populating data with vertices_to and morph_mat I would do it
with at least a private method of SourceMorph class.

Basically you want to isolate as much as possible the morphing logic inside the SourceMorph class.

TommyClausner replied:
stop reading my mind! :P

TommyClausner commented:
at the moment I'm sitting there and try to spot all those things anyway, after all tests are passed... So maybe it would be good to wait until I fixed the stuff I spotted by myself, so that we avoid double reviewing(?)

agramfort commented:
I'll do another pass when you tell me to

TommyClausner commented:
Cool, thanks! ;)

agramfort commented:
Great minds think alike ...

TommyClausner commented:
all right, let's see how alike they are ;)

I think there might be still a lot of things that I would have liked to change, but didn't "see" it, so I guess a second look is now really appreciated :)

TommyClausner commented:
btw I have some sketch of a tutorial in the pipe... I just thought keeping it back would maybe not overload the PR too much, but on the other hand I'm not sure if a few more lines would make it so much worse ;)  - so if you like I could also push it(?) - Some question regarding the style of a tutorial: I made it now such that I present + explain most of it's functionality and the parameters + Have a "in a nutshell" section where everything is condensed.

The only thing is, that it might compute too long (1.5 - 2 min) if I want to show all functionality. So I thought I could explain it and make "text examples" and only compute one real example? Or shall I compute some different variations and compare them via plot or so?

agramfort commented:
regarding timing if it passes on circleci I would say it's acceptable.

did you update the existing tutorial? We should consolidate and remove what becomes obsolete.

agramfort commented on mne/source_estimate.py:
```diff
@@ -2454,236 +2249,31 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
+    _update_morph_data(source_morph, {'0': stc_from.lh_vertno,
+                                      '1': stc_from.rh_vertno,
+                                      'hemis': [0, 1]},
+                       kind='surface')
```
I would do:

`source_morph._update_morph_data_surf(stc_from)`

agramfort commented on mne/source_estimate.py:
```diff
@@ -2454,236 +2249,31 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
+                       kind='surface')
 
+    # compute morph
+    _compute_morph_data(source_morph, verbose=verbose)
```
`source_morph._compute_morph_data(verbose=verbose)`

agramfort commented on mne/source_estimate.py:
```diff
@@ -2692,22 +2282,23 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
     stc_to : SourceEstimate | VectorSourceEstimate
         Source estimate for the destination subject.
     """
+    from mne.morph import SourceMorph, _update_morph_data
```
otherwise you have a circular import?

agramfort commented on mne/source_estimate.py:
```diff
@@ -2725,20 +2316,16 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
+    # use pre-computed data
+    _update_morph_data(source_morph,
+                       {'vertno': vertices_to, 'morph_mat': morph_mat},
+                       'surface')
```
source_morph._update_morph_data(...)

TommyClausner commented:
Regarding the tut: I made a new one, that is actually demonstrating all aspects of SourceMorph - so you'd prefer updating old ones + a new one showing features that are not shown in others?

agramfort commented:
@TommyClausner you opened again a PR from master... I have a commit for you:

https://github.com/agramfort/mne-python/commit/ed3f15735d7966010356e568fc3691788b81a660

I don't want to push it brutally to your master. To get it do:

```
git remote add agramfort https://github.com/agramfort/mne-python.git
git fetch agramfort
git cherry-pick ed3f15735d7966010356e568fc3691788b81a660
```

let me know if you need help or we should discuss.

TommyClausner commented:
The low code cov seems to come from the slowtest flags... testing ``morph.py`` with ``test_morph.py``, ``test_source_estimate.py`` and ``test_label.py`` yields 91%

larsoner commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,258 @@
+
+    # check as_volume=True
+    assert isinstance(source_morph_vol(stc_vol, as_volume=True),
+                      nib.Nifti1Image)
```
We should probably change to `Nifti2Image` as the default output in these new methods, see #5337

agramfort replied:
it's an overkill to produce a volume to test output type.
Test type next to the line where you test this for correct output dim or whatever else you're testing.
THM -> try to be smart to avoid as much as possible testing time.

TommyClausner commented:
I'll add the option to #5338 :)

agramfort commented:
@TommyClausner I see above in your todos a few remaining items:
- account for mixed spaces -> I think we should do it in a next PR
- tutorial. What's the status here?

how long is the new testing time? and you reduce it more?

TommyClausner commented:
@agramfort you are right - guess it's a good idea to make a new pr for mixed
The tutorial is actually ready(-ish) - depends a bit, because I'm not sure if how I did it is the way you would like it... Shall just push it to this pr?

agramfort commented:
yes push the tutorial.

the page should be generated with sphinx gallery so the content
comes with code that we know works.

it means that the main entry point for morphing that currently points to:

http://martinos.org/mne/stable/manual/source_localization/morph.html

should change and this content should be integrated.

sounds good?

TommyClausner commented:
all right I'll push soon - just to make sure I got it right: the stuff from morph.rst goes into the tutorial I made or vice versa and everything goes in morph.rst (guess it's .rst)?

agramfort commented:
option 1) everything goes into a plot_morph.py tutorial.
option 2) you write a plot_morph.py tutorial and you make an advanced page
based on morph.py and plot_morph.py points to it

TommyClausner commented:
I think the second option would be nicer somehow. Because the problem with the tutorial as it is atm is, that I try to show everything, which either causes a lot of commented code or a lot of computation time - so I think I could make the tutorial for a single use case for both vol and surf and add all description of functionality to an "advanced page"?

larsoner commented:
> But it's not really nice to throw all of them in a single example / tutorial

That is my thought, we can just update existing tutorials that use `as_volume` and `stc.morph` etc. and people will see the the methods where they are relevant. Hopefully this will already show the ways in which your methods are nicer.

TommyClausner commented:
I updated 3 existing that I found and just technically was letting them as they were, just using the new API. The problem I see is, that on one hand it would be cool to have somehow a guided explanation of the full functionality - so this would be nice to have in a file somehow (just not tut or ex) - because when I add some functions to the existing tutorials randomly (to expose all) it's really confusing to the user I guess

larsoner commented:
This can go in the Notes of the object and its methods

>

agramfort commented:
can you share a link to rendered tutorial on circle when you have it?

TommyClausner commented:
tut: https://8628-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_morph.html
ex: https://8628-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_use_sourcemorph.html

TommyClausner commented:
thanks @agramfort that's why I called it a sketch ;)

ok yeah, sounds good what you suggest - I then go for ``plot_morph.py`` to demonstrate what I have in the "in a nutshell section and ``WhateverAdvancedExample.Whatever`` for the full functionality(?) - I'll also only demonstrate on the volume in the tutorial, right (because surf is actually covered)?

agramfort commented:
I would do something like:

- what is morphing and what is it useful for.
- how does it work conceptually with surfaces and volumes (you can take
text from here:
https://www.frontiersin.org/articles/10.3389/fnins.2013.00267/full for
surface case)

then how does it work in practice with mne python making plots that help
understand the concepts.

TommyClausner commented:
All right, I already spotted quite some typos in the tutorials (beside those Travis found ;) ), so I guess I'll work tomorrow on that... Beside typos there were some assertion errors:

https://travis-ci.org/mne-tools/mne-python/jobs/407759601
2717, 2927, 3015

Is it possible to tell if they are related to what I did?

Other than that, here are the renderings:
Example: https://8664-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_use_sourcemorph.html

Short tutorial: https://8664-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_morph.html

Long tutorial: https://8664-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_morph.html

Especially for the long tutorial: Would you think it's (content wise) fine-ish - I know I have to correct some language mistakes and see that I'm consistent in how I call things; But beside this? Shall I go more into detail in the volume morph theory? Because the math is quite heavy and I'm not sure how much sense it makes to put it there... I tried to shortly rephrase it in kinda simple, but (hopefully) informative words and referred to the paper. So I could extend this if you like by some graphic or morph a "checker volume" to show how the mapping looks like or so?

jasmainak commented:
Two comments from a superficial look on the long tutorial:

1. The parameters of the function don't need to be detailed in my opinion
2. The references and many other links are not rendered properly

Also yeah, I wouldn't put too much math in there. It's a tutorial -- so it shouldn't scare away people :-)

agramfort commented on examples/inverse/plot_use_sourcemorph.py:
```diff
@@ -0,0 +1,101 @@
+
+# Obtain absolute value for plotting
+# To not copy the data into a new memory location, out=stc.data is set
+np.abs(stc.data, out=stc.data)
```
I don't think you need it here as dSPM output if the norm by default so it's already >= 0

agramfort commented:
what is the difference between:

https://8664-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_use_sourcemorph.html

and this:

https://8664-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_morph.html#step-by-step-hands-on-tutorial

?

I agree with @jasmainak. Don't duplicate the docstring of the SourceMorph
object by explaining all parameters. Explain the concepts first and explain
what 90% of the users will want to use.

I don't have more time to comment line by line sorry.

TommyClausner commented:
Hey thanks so far! 

I explained the parameters so detailed, to not just point out what they mean and do, but also explain it and relate it to the theory... At least this was the idea (kinda to have a collection of info that someone could look up, but that would probably be too much for a docstring in the class)... Hm so completely removing it?

@agramfort I guess you mean the difference between the second you posted and the other (short tutorial not the one you posted as first) - the difference is that the short tutorial is a hybrid between the example (more information than that + some explanation) and the long tutorial version (less info; stripping details, that probably 80-90% wouldn't care too much about (like optimization parameters))... if you wish I also could delete the short tutorial and refer to the "hands-on-tutorial" in the long tutorial directly(?) - again: my idea was actually to make a tutorial for those who just wanna do it and those who care about the details + the example to just have one

TommyClausner commented:
Hey :) - ok now most typos are hopefully gone and everything is a bit more consistent... - I removed the doc-string-like description for the parameters but kept it for the methods. So far I'm also still in favor of keeping both tutorials :) - btw: I added some ok-ish math that I think clarify things more than it adds confusion... - renderings should be fine now as well (hope)

TommyClausner commented:
and the renderings:

Example:
https://8676-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_use_sourcemorph.html

Short Tutorial:
https://8676-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_morph.html

Background Info:
https://8676-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_morph.html#sourcemorph

jasmainak commented on tutorials/plot_background_morph.py:
```diff
@@ -0,0 +1,875 @@
+Nevertheless, this can be a first step of a successful subject to subject
+transformation.
+
+We can understand mapping as a mathematical function :math:`f`. This function
```
not rendered correctly. You may want to build the docs locally and check first before pushing ...

jasmainak commented on tutorials/plot_background_morph.py:
The last two paragraphs feel like overly verbose to me. In my opinion, we should simply point to the wikipedia page for affine transformations: https://en.wikipedia.org/wiki/Transformation_matrix#Affine_transformations

TommyClausner replied:
... ok :(

But do you think it's obvious to people who never thought about transformation matrices, why translation actually requires an additional dimension (I just really like the way of expressing this as shearing in higher dimension)? Also: wikipedia is also not so nice and clear about that ;)

removing?

jasmainak commented on tutorials/plot_background_morph.py:
```diff
@@ -0,0 +1,875 @@
+    y \\
+    z \\
+    1
+    \end{bmatrix}\ ,
```
does this mean `M^{(AB)} = I_4` ?

TommyClausner replied:
Err... yes... if I_4 means identity matrix with 4 columns then yes... but why? Shall I indicate that? I just used the identity matrix when I wrote it out to get a visual depiction of some transformation matrix. It also nice to make this easy example with scaling, because it becomes immediately obvious how transformation matrices work (I think ;) )

jasmainak commented on tutorials/plot_background_morph.py:
link here: https://martinos.org/mne/dev/manual/c_reference.html#mne-make-morph-maps ?

jasmainak commented on tutorials/plot_background_morph.py:
non-zero

jasmainak commented on tutorials/plot_background_morph.py:
abuse of notation? if you index `x^{(B)}` from above, you will get a single coordinate. But that is not what you mean, right?

jasmainak commented on tutorials/plot_background_morph.py:
same problem here ... the dimensions of `M^{(AB)}` have changed, right?

jasmainak commented on tutorials/plot_background_morph.py:
now what is `x_{(A)}` ?

jasmainak commented on tutorials/plot_background_morph.py:
and is `x_{(A)}` different from `x^{(A)}`. Is one of them estimated whereas the other the true value?

jasmainak commented on tutorials/plot_background_morph.py:
you mean the total amplitude?

jasmainak commented on tutorials/plot_background_morph.py:
indices of non-zero

jasmainak commented on tutorials/plot_background_morph.py:
Then the composite blurring matrix is a detour that is not relevant for the reader and should be removed.

jasmainak commented on tutorials/plot_background_morph.py:
what does this have to do with morphing? Isn't this done whenever we visualize the `stc` results? anything specific to morphing here?

jasmainak commented on tutorials/plot_background_morph.py:
step-by-step

jasmainak commented on tutorials/plot_background_morph.py:
provide weblinks?

jasmainak commented on tutorials/plot_background_morph.py:
This is a nice mathematical fact, but is it relevant to understand the tutorial?

TommyClausner replied:
yeah I know it's kinda unrelated - my goal was to get someone who maybe didn't think about this before just some grasp of what can be expected in terms of how different the data could be for both representations etc - removing?

TommyClausner commented:
Hey @jasmainak thanks for all the comments!
here is the freshly rendered version that at least got rid of some rendering issues:
https://8678-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_morph.html

Now, sorry I should have probably said something about what I copied, because actually many of your comments refer to the surface morph text, which I copied from ``morph.rst`` So at least the sections:

Surface morphing

    Morph maps
    Smoothing
    Precomputing

are actually a (maybe even literally) a one to one copy from that file. The idea was - and I discussed this shortly with @agramfort if I understood correctly - to delete ``morph.rst`` instead. Further I didn't "check" those section - I only made sure that they are rendered correctly etc. 

Regarding the math there: honestly I found the notation of x^(B) and M^(AB) etc kinda weird, and I think it's not the best way of expressing the math there (at least it doesn't seem too common). Before I had it differently in my examples like using "a" and "b" directly as vectors or even more abstract using " d " and " d' " . But then I've seen how the notation was in the surface section and adjusted myself.

I'll "ignore" the surface comments for now and reply to some other first, so that maybe more people can decide if we remove ``morph.rst`` and what the information about surfaces should be etc. (Also would happily change to the mathematical notation there ;) )

TommyClausner commented:
All hey again @jasmainak 

All your comments I did not answer yet were either typos or related to the surface morph from ``morph.rst`` that I copied (copied literally everything in there, because it was called "morph").

I would now try to make the tutorial a bit more fluent to read and maybe less verbose (Even though I really like the info there... it's informative :)  ) - but what shall we do about the surface section? shall I completely revise it or can it stay because the person who wrote it has reasons why he/she made it this way (there was no author)?

TommyClausner commented:
Hey there!

short question: do I have to do something specifically to make the docstring of __call__ being rendered? No matter what I try it won't appear in the class doc and I cannot reference it. How do I do that??

larsoner commented:
It should happen automatically because of 

https://github.com/mne-tools/mne-python/blob/master/mne/utils.py#L73

... but it looks like someone (probably me!) did not heed the warning there, it also needs to be added here:

https://github.com/mne-tools/mne-python/blob/master/doc/_templates/class.rst

Can you make sure the lists are identical?

TommyClausner commented:
really?? I tried exactly this and it didn't render either

TommyClausner commented:
oh man! Be praised @larsoner it worked! (cleaning helped indeed #facepalm)

TommyClausner commented:
Here we go:

Example:
https://8693-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_use_sourcemorph.html

Short tutorial:
https://8693-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_morph.html

Long tutorial:
https://8693-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_morph.html

jasmainak commented:
@TommyClausner don't duplicate the documentation. You should delete somewhere if you add somewhere. Otherwise it is really hard to know what's going on. Nobody (except maybe @agramfort and @larsoner ) are familiar with the whole codebase. Make life easier for the code reviewer :-)

jasmainak commented:
@TommyClausner your short tutorial and long tutorial are duplicating code, right? We should avoid that ...

Let's try to avoid as much duplication as possible.

agramfort commented:
@TommyClausner <https://github.com/TommyClausner> I agree with @jasmainak.
We should avoid duplicating content and be better in pedagogy.

let me try to help:

- rename plot_use_sourcemorph.py to plot_morph_volume_stc.py
- create the equivalent one for surface called plot_morph_surface_stc.py

in the plot_morph.py explain:

- why do we need morphing? how does it work? ie based on anatomical
warpping.
- explain conceptually how it works for surface stc. And point to
plot_morph_surface_stc.py
- explain conceptually how it works for volume stc. And point to
plot_morph_volume_stc.py

everything that is left should be put in plot_morph_advanced.py to which
you point at the end of plot_morph.py

ok?

TommyClausner commented:
Hey @jasmainak and @agramfort 

all right, let me rephrase in order to avoid confusion (on my end ;) ):

- separate long tutorial from this PR(?)
- make separate examples for surface and volume
- basic tutorial: general theory (from long tut?), conceptual surf + vol using SourceMorph + important differences, link to examples
- advanced tutorial: (no general theory?), theory about surf (from morph.rst)+ vol, details about SourceMorph

correct?

jasmainak commented:
yep, separate long tutorial. Make separate examples for surf and volume. Not sure you need to keep any of the stuff from the long tutorial for now. I'd leave out the theory part for now. Just focus on the code and exposing the API with one-line explanations where necessary. Basically, no new example/documentation file should exceed 100 lines.

agramfort commented:
yes. Keep examples short.

Keep the maths and advanced stuff in a an advanced file / section.

Imagine there is a master student with no background in M/EEG that comes
to you and you need to explain what is it useful for. What is needed to
make it work.
What is needed as input and what will be the output. How do you look that
it went fine etc.

TommyClausner commented:
Hey there :)
after having won against sphinx (hopefully... it's a sneaky little thing...) I finally made it...
I post the links once rendered:


Now it's such, that there is one example for vol and one for surf + the tutorial (which actually could become ``morph_stc.rst`` instead of ``plot_morph_stc.py``).

I tried to make it beginner friendly and not tooooo long. In order to avoid doubling code, I just now use the result plots from the examples and show them in the tutorial and reference to the example in order to know how they were computed.

There is still some code, but I used >>> rather than actual code... So now there is actually no code left that's executed twice.

TommyClausner commented:
Hey :)

I was playing around a little bit with mixed source estimates and would have some questions:
- is it guaranteed that there are always only two surfaces that make the first two list entries of src?
- if not: are surfaces that make two hemispheres are always consecutive in src?
- is there a difference between multiple volumetric (but also surface) spaces except for the data interpolation? Or in other words: do I always only have to compute one volumetric and one surface morph for all spaces in src or one for each space?

Thanks a lot !

Cheers,
Tommy

jasmainak commented:
> I tried to make it beginner friendly and not tooooo long. In order to avoid doubling code, I just now use the result plots from the examples and show them in the tutorial and reference to the example in order to know how they were computed.

oh yesss! we're getting there. I do another round of code review later in the day

jasmainak commented:
@TommyClausner how long do the tests take now? You had a timeout on one build of CircleCI

TommyClausner commented:
Hey @jasmainak here are the slowest SourceMorph related tests:

15.17s call     mne/tests/test_source_estimate.py::test_morph_data
13.83s call     mne/tests/test_morph.py::test_volume_source_morph
4.45s call     mne/tests/test_morph.py::test_surface_vector_source_morph
3.61s call     mne/tests/test_source_estimate.py::test_stc_methods
1.95s call     mne/tests/test_label.py::test_morph
0.52s call     mne/tests/test_source_estimate.py::test_save_vol_stc_as_nifti
0.44s call     mne/tests/test_morph.py::test_stc_as_volume
...

jasmainak commented:
So you simply split the 30 seconds worth of testing into 3 tests? No way to make it more efficient as I suggested earlier?

jasmainak commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,272 @@
+        subject_from='sample',
+        subject_to=os.path.join(data_path, 'subjects', 'fsaverage', 'mri',
+                                'brain.mgz'),
+        subjects_dir=subjects_dir, niter_affine=(10, 10),
```
use fewer iterations? do you actually check the quality of morphing?

jasmainak replied:
you could even do just 1 or 2 iterations ...

jasmainak commented on tutorials/plot_background_filtering.py:
```diff
@@ -929,7 +929,7 @@ def baseline_plot(x):
 #
 # Filter length (order) and transition bandwidth (roll-off)
-# --------------------------------------------------------
+# ---------------------------------------------------------
```
Try to avoid cosmetic changes. See point 4 [here](http://blog.ploeh.dk/2015/01/15/10-tips-for-better-pull-requests/).

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    ----------
+    subject_from : str | None
+        Name of the original subject as named in the SUBJECTS_DIR.
+        If None src[0]['subject_his_id]' will be used. The default is None.
```
should be `src[0]['subject_his_id']`

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        If None src[0]['subject_his_id]' will be used. The default is None.
+    subject_to : str | array | list of two arrays
+        Name of the subject on which to morph as named in the SUBJECTS_DIR
+        The default is 'fsaverage'. If morphing a surface source extimate,
```
surface source estimate

jasmainak commented on mne/source_estimate.py:
```diff
@@ -2454,260 +2257,42 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
+                       warn=warn, verbose=verbose)(stc_from)
 
 
+@deprecated("This function is deprecated and might be removed in a future "
```
will be removed :-)

massich replied:
when is deprecated and when would it be removed?

massich replied:
deprecated in 17, removed in 19

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+
+###############################################################################
+# Morph for SourceEstimate |  VectorSourceEstimate
+@deprecated("This function is deprecated and might be removed in a future "
```
The deprecated functions need to be removed from `python_reference.rst`, no? Otherwise, new users will keep going to these old functions

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        If spacing is None, MRIs won't be resliced. Note that in this case
+        both volumes (used to compute the morph) must have the same number of
+        spatial dimensions.
+        If morphing SourceEstimate or Vector SourceEstimate, spacing can be
```
VectorSourceEstimate

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        ``subject_to == subject_from``. See notes below.
+    precomputed : dict | None
+        Precomputed morphing data. For (Vector)SourceEstimates it contains
+        of 'morph_mat' and 'vertno' where morph_mat is the respective
```
of 'morph_mat' -> 'morph_mat'

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        transformation and vertno the corresponding vertices. For
+        VolSourceEstimates it should contain 'DiffeomorphicMap' and
+        'AffineMap' data, as well as 'morph_shape', 'morph_zooms' and
+        'morph_affine', referring to the respective volume parameters.
```
This is too much info. Just say how you can get this `precomputed` dict. Is there a particular function to call?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        Morph across hemisphere.
+    params : dict
+        The morph data. Contains all data relevant for morphing and / or volume
+        output.
```
Since you have a `repr now`, you don't need all these attributes to be public, right? If you have public attributes, users may attempt to fiddle with them which can lead to unexpected results ... For the ones that you really need to expose, you can use `@property` for them

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+        mri_resolution: bool | tuple | int | float
+            If True the image is saved in MRI resolution. Default False.
+            WARNING: if you have many time points the file produced can be
+            huge.
```
better to put this warning in the `save` method, no?

TommyClausner replied:
I actually wouldn't like to put it to the save method, because it will become huge here, not just when saving (it's huge also in working memory)... ?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+            self._update_morph_data(_get_src_data(src))
+            self._compute_morph_data(verbose=verbose)
+
+    def __call__(self, stc_from, as_volume=False, mri_resolution=False,
```
you're missing `@verbose` decorator

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+            If not None, override default verbose level (see
+            :func:`mne.verbose` and
+            :ref:`Logging documentation <tut_logging>`
+            for more). Defaults to self.verbose.
```
I don't see `self.verbose` is ever assigned

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    verbose : bool | str | int | None
+        If not None, override default verbose level (see
+        :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+        for more). Defaults to self.verbose.
```
what is `self` here?

jasmainak commented on examples/inverse/plot_lcmv_beamformer_volume.py:
```diff
@@ -93,7 +93,7 @@
 
 # take absolute values for plotting
-stc.data[:, :] = np.abs(stc.data)
+np.abs(stc.data, out=stc.data)
```
again, unrelated change ...

agramfort replied:
I can live with it. I am getting old and more permissive... :)

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+plt.show()
+
+# Save memory
+del stc, stc_fsaverage, inverse_operator, morph
```
is it really useful to delete these objects after having run everything? It prevents me from trying stuff on ipython since the objects are no longer in memory

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+morph = SourceMorph(subject_from='sample',
+                    subject_to='fsaverage',
+                    subjects_dir=subjects_dir,
+                    src=inverse_operator['src'])
```
I get a bunch of deprecated warnings like this one:

```py
:985: RuntimeWarning: invalid value encountered in sqrt
  mean_norm = np.sqrt(sq_field).mean()
```

Why is this happening?

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+
+# Plot glass brain (change to plot_anat to display an overlaid anatomical T1)
+display = plot_glass_brain(t1_fsaverage, display_mode='ortho',
+                           cut_coords=[0., 0., 0.],
```
In `plot_glass_brain`, the `cut_coords` does nothing if I remember correctly ...

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+fig.patch.set_facecolor('white')
+
+# Plot glass brain (change to plot_anat to display an overlaid anatomical T1)
+display = plot_glass_brain(t1_fsaverage, display_mode='ortho',
```
`display_mode='ortho'` by default in this function. No need to advertise it more.

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+                           cut_coords=[0., 0., 0.],
+                           draw_cross=False,
+                           axes=axes,
+                           figure=fig,
```
why do we need to pass figure and axes handle. [Keep it Simple Stupid](https://en.wikipedia.org/wiki/KISS_principle).

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+
+# Add annotations and title
+display.annotate(size=8)
+axes.set_title('subject results to fsaverage', color='black', fontsize=12)
```
you can pass the `title` to plot_glass_brain

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+# Add annotations and title
+display.annotate(size=8)
+axes.set_title('subject results to fsaverage', color='black', fontsize=12)
+plt.text(plt.xlim()[1], plt.ylim()[0], 't = 0.087s', color='black')
```
why is this t=0.087?

jasmainak commented on examples/inverse/plot_morph_surface_stc.py:
```diff
@@ -0,0 +1,72 @@
+brain.add_text(0.1, 0.9, 'Morphed to fsaverage', 'title', font_size=16)
+
+# Save memory
+del stc, stc_fsaverage, morph
```
same comment here. Let's not delete ...

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    n_vertices_seen = 0
+    for k, v in enumerate(vol):  # loop over time instants
+        stc_slice = slice(n_vertices_seen, n_vertices_seen + n_vertices)
+        v[mask3d] = stc.data[stc_slice, k]
```
you're modifying the variable you are looping over????!

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    n_vertices = np.sum(mask3d)
+
+    n_vertices_seen = 0
+    for k, v in enumerate(vol):  # loop over time instants
```
Single letter variable names are considered bad for readability. You can't search for them using Ctrl + F. And also, they are not descriptive

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+MRIs to use, ``subject_from`` and ``subject_to`` need to be defined as the
+name of the respective folder in the FreeSurfer's home directory.
+
+subject_from can also be inferred from src, subject_to set is to 'fsaverage'
```
``subject_from``

jasmainak replied:
That is, you need to add the quotes to format it properly

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+--------------------------------------------
+
+We can convert our morphed source estimate into a NIfTI volume using
+morph.as_volume(). We provided our morphed source estimate as first argument.
```
need to format properly

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+All following keyword arguments can be used to modify the output image.
+
+See :meth:`morph.as_volume() <mne.SourceMorph.as_volume>` for more
+information.
```
superfluous sentence

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+See :meth:`morph.as_volume() <mne.SourceMorph.as_volume>` for more
+information.
+
+Note that apply_morph=False, that is the morph will no be applied because the
```
Formatting of ``apply_morph=False``

jasmainak replied:
will no -> will not

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+    >>> morph = read_source_morph('my-file-name-morph.h5')
+
+Shortcuts
+=========
```
I'm a bit scared of this section. According to the Zen of Python:

> There should be one-- and preferably only one --obvious way to do it.

but I suppose you're not Dutch :)

agramfort replied:
I agree with @jasmainak . Is it really just a short cut ? maybe is allows to get a volume without morphing ie staying in the same subject?

jasmainak commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+r"""
+===========================================
+Morphing source estimates using SourceMorph
+===========================================
```
The new tutorial reads great. It's simple and easy to follow! But see my comment about last section

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+            The source estimate to morph.
+        as_volume : bool
+            Whether to output a NIfTI volume. stc_from has to be a
+            VolSourceEstimate. Default is False.
```
do we really need this `param`?

jasmainak commented on examples/inverse/plot_morph_volume_stc.py:
```diff
@@ -0,0 +1,105 @@
+###############################################################################
+# Morph VolSourceEstimate
+
+# Initialize SourceMorph for VolSourceEstimate
```
@agramfort I don't know what you feel about this API. But it took me a while to realize that `SourceMorph()` is actually doing some computation, not just creating an object.

jasmainak replied:
I kept wondering at what step the morphing matrix was computed

agramfort replied:
yes it's done in the constructor of SourceMorph()

I think it's the most natural way to do it.

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+
+
+def _compute_morph_data(morph, verbose=None):
+    """Compute source estimate specific morph."""
```
this does not use the source estimate in any way, does it?

jasmainak commented:
@TommyClausner that's it from me for today. Sorry I took so long to get to this. You can update the todo list and address my comments. I think we're getting close!!! After this, we can let @agramfort and @larsoner take a look

TommyClausner commented:
Hey together!

Thanks for all the feedback! Especially @jasmainak for taking the time for a detailed view!

I'll work it in - some things I'm not totally agreeing on (e.g. with removing the attributes, because I think the user should be allowed to mess up his / her pipeline if he / she wishes - at least I appreciate that usually) - but I'll just work the stuff in (leaving out the things I'm not agreeing on) and when I push it comment on those (few) things and we can finally discuss them ;)

PS: sorry for not being responsive yesterday (one of those days that force you to a lot of male cattle excrement taking the whole day for nothing :/ - I'll catch up)

massich commented on examples/inverse/plot_morph_surface_stc.py:
```diff
@@ -0,0 +1,72 @@
+Morph surface source estimate
+=============================
+
+This example demonstrates how to morph an individual subject **surface source
```
I would rephrase this:

massich commented on examples/inverse/plot_morph_surface_stc.py:
```diff
@@ -0,0 +1,72 @@
+=============================
+
+This example demonstrates how to morph an individual subject **surface source
+estimate** to a common reference space. For this purpose we will use
```
This example demonstrates how to morph an individual subject **surface source
￼estimate** to a common reference space.

massich replied:
I would rephrase it in this manner:

massich commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    header.set_xyzt_units('mm', 'msec')
+    header['pixdim'][4] = 1e3 * stc.tstep
+
+    with warnings.catch_warnings(record=True):  # nibabel<->numpy warning
```
this catches all warnings, maybe you want to check for this `nibabel<->numpy warning` and let the rest pass.

I'm not sure if there's a way to do something like this:
```py
with warnings.catch_warnings():
     warnings.simplefilter('ignore', XXXWarning)
     img = NiftiImage(vol, affine, header=header)
```
but that it also allows for some regex, to make sure that you catch what you want but not somethings else.

massich commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    This function will be used to morph VolSourceEstimate based on an
+    affine transformation and a nonlinear morph, estimated based on
+    respective transformation from the subject's anatomical T1 (brain) to
+    a destination subject's anatomical T1 (e.g. fsaverage). Afterwards the
```
can we say reference instead of destination?

massich commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+    respective transformation from the subject's anatomical T1 (brain) to
+    a destination subject's anatomical T1 (e.g. fsaverage). Afterwards the
+    transformation can be applied. Affine transformations are computed
+    based on the mutual information. This metric relates structural changes
```
reword this part
```
This metric relates structural changes
    in image intensity values. Because different still brains expose high
    structural similarities this method works quite well to relate
    corresponding features [1]_.
```

massich commented on mne/morph.py:
```diff
@@ -0,0 +1,1443 @@
+                                                   mri_from,
+                                                   mri_from_grid2world)
+
+    nbins = 32
```
why is this hardcoded like this? 
if this has to be the case I would do this:

`imaffine.MutualInformationMetric(nbins=32, whatever=None)`

massich commented on mne/source_estimate.py:
```diff
@@ -1790,16 +1787,16 @@ def save_as_volume(self, fname, src, dest='mri', mri_resolution=False):
-        mri_resolution: bool
-            It True the image is saved in MRI resolution.
+        mri_resolution : bool
+            Whether to use MRI resolution.
```
Whether to use MRI resolution when saving

massich commented on mne/source_estimate.py:
```diff
@@ -1840,8 +1840,11 @@ def as_volume(self, src, dest='mri', mri_resolution=False):
         """
-        return _save_stc_as_volume(None, self, src, dest=dest,
-                                   mri_resolution=mri_resolution)
+        return _save_stc_as_volume(None, self,
```
I find the previous more readable.

massich commented on mne/source_estimate.py:
```diff
@@ -3093,8 +2667,8 @@ def save_stc_as_volume(fname, stc, src, dest='mri', mri_resolution=False):
-    mri_resolution: bool
-        It True the image is saved in MRI resolution.
+    mri_resolution : bool
+        Whether to use MRI resolution
```
idem

massich commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,272 @@
+fname_t1 = os.path.join(subjects_dir, 'sample', 'mri', 'T1.mgz')
+
+
+def _real_vec_stc():
```
+1

massich commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,272 @@
+@sample.requires_sample_data
+def test_volume_source_morph():
+    """Test volume source estimate morph, special cases and exceptions."""
+    import nibabel as nib
```
why nib? nibabel is pretty enough.

massich replied:
or `from nibalbel import Nifti1Image, Nifti2Image`

agramfort replied:
we use nib everywhere.

Its also the standard import name of this lib: http://nipy.org/nibabel/gettingstarted.html

massich commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,272 @@
+        niter_sdr=(3, 3), spacing=7)
+
+    # check wrong subject_to
+    pytest.raises(IOError, SourceMorph,
```
use the context, it is WAY more readable. + check that you raise what you want.
```py
with pytest.raises(IOError, match='something meaninful'):
    SourceMorph(subject_from='sample', subject_to='42' ..
```

massich commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,272 @@
+    img_any_res = source_morph_vol.as_volume(stc_vol_morphed,
+                                             mri_resolution=(5., 5., 5.),
+                                             fname=os.path.join(tempdir, '42'))
+    assert isinstance(img_any_res, nib.Nifti1Image)
```
assert isinstance(img_any_res, Nifti1Image)

massich commented on mne/tests/test_source_estimate.py:
```diff
@@ -543,122 +547,149 @@ def test_extract_label_time_course():
+    assert sum('deprecated' in str(ww.message) for ww in w) == 2
+
+    with warnings.catch_warnings(record=True):
+        pytest.raises(ValueError, stc_from.morph_precomputed,
```
use context aswell

agramfort replied:
@massich can you clarify what you mean?

massich commented on tutorials/plot_morph_stc.py:
```diff
@@ -0,0 +1,248 @@
+to a common space.
+
+We will use precomputed data and morph surface and volume source estimates to a
+common space. The common space of choice will be FreeSurfer's "fsaverage".
```
For this particular example 'fsaverage' from FreeSurfer is used as reference space.

I like reference better than common space. (but whatever it is, it should be consistent with the docstrings)

massich commented on examples/inverse/plot_morph_surface_stc.py:
```diff
@@ -0,0 +1,72 @@
+=============================
+
+This example demonstrates how to morph an individual subject **surface source
+estimate** to a common reference space. For this purpose we will use
```
maybe `source estimate` is a candidate to go to glossary #5395

agramfort commented:
+1

TommyClausner commented:
Hey :)

pew... here we go (I'll post the links when rendered):

<links>

I implemented most of the suggestions, but would like to discuss some of them:

- I kept the "cosmetic changes", since they are now in there anyway... I will avoid that in the future, ok?

- removed the deletion of working memory items (found the argument good, that I maybe want to try multiple settings or so and want it in WM)

- de-deprecated ``compute_morph_matrix`` because it's referenced all across. So maybe this particular one can go a cycle later or never, since we also have now ``precomputed=``- "shall it stay or shall it go?"

- I ignored some style things that were mentioned in order to keep things consistent with previous stuff

- kept public API because I like it way more (can be discussed if desired ;) )

- did not modify ``check_version`` yet. Shall I make ``check_version(..., raises=bool)``?

- kept nibabel as nib, because it's actually quite common... same would be to argue that numpy is also fine just as is... ;) - change nib to nibabel or keep?

- kept the "Shortcuts" section in the tut... I mean I'm a punk doing this stuff anyway, so we could also be transparent with the user... no?

It was quite hard to find a way to explain ``SourceEstimate`` and FreeSurfer, because I'm not too familiar with the full functionality of those... So it could be that my naive view made me focus on things you would like to remove and forgot things you might find important...

TommyClausner commented:
no CircleCI?

larsoner commented:
Did you have CircleCI enabled for your fork and then disable it? We've had problems with CircleCI picking back up for `mne-tools` when users have done that before.

agramfort commented on doc/documentation.rst:
```diff
@@ -112,6 +112,7 @@ There are also **examples**, which contain a short use-case to highlight MNE-fun
     auto_tutorials/plot_object_raw.rst
     auto_tutorials/plot_object_epochs.rst
     auto_tutorials/plot_object_evoked.rst
+    auto_tutorials/plot_object_sourceestimate.rst
```
object is called SourceEstimate so in snake case it should be source_estimate

agramfort commented on doc/documentation.rst:
```diff
@@ -431,6 +434,8 @@ There are also **examples**, which contain a short use-case to highlight MNE-fun
     auto_examples/inverse/plot_mne_crosstalk_function.rst
     auto_examples/inverse/plot_mne_point_spread_function.rst
     auto_examples/inverse/plot_morph_data.rst
```
do we need to keep plot_morph_data.py ?

agramfort commented on doc/whats_new.rst:
```diff
@@ -19,6 +19,8 @@ Current
+
 - Add possibility to save :class:`mne.VolSourceEstimate` and :class:`mne.MixedSourceEstimate` to HDF5 format (file extension .h5) with :meth:`mne.VolSourceEstimate.save` and :meth:`mne.MixedSourceEstimate.save` by `Alex Gramfort`_
 
 - Add ability to pass ``axes`` to ``ts_args`` and ``topomap_args`` of :meth:`mne.viz.plot_evoked_joint` by `Jona Sassenhagen`_
```
@TommyClausner you need to add in API section all deprecated functions.

agramfort commented on examples/inverse/plot_morph_surface_stc.py:
```diff
@@ -0,0 +1,70 @@
+This example demonstrates how to morph an individual subject
+:class:`mne.SourceEstimate` to a common reference space. For this purpose we
+will use :class:`mne.SourceMorph`. Pre-computed data will be morphed based on
+an affine transformation on the surface vertices towards the reference
```
it's not affine. It's based on spherical registration of FreeSurfer.

agramfort commented on mne/tests/test_label.py:
```diff
@@ -692,6 +692,8 @@ def test_stc_to_label():
 
 
+@requires_nibabel()
+@requires_dipy()
```
a test that did need dipy now needs it? this will break code. You should avoid this.

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+    assert isinstance(source_morph_surf, SourceMorph)
+
+    assert isinstance(SourceMorph(subjects_dir=subjects_dir)(stc_surf),
+                      SourceEstimate)
```
these 2 asserts are not necessary.

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+
+    assert (all([read == saved for read, saved in
+                 zip(sorted(source_morph_surf_r.__dict__),
+                     sorted(source_morph_surf.__dict__))]))
```
you are only checking keys and not values. Is it intentional?

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+    import nibabel as nib
+    data_path = sample.data_path(
+        download=False)  # because testing data has no brain.mgz
+    subjects_dir = os.path.join(data_path, 'subjects')
```
os.path -> op?

you changed this above.

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+    #     os.path.join(sample_dir, 'sample_audvis-meg-vol-7-fwd.fif'),
+    #     subject_from='sample',
+    #     subjects_dir=os.path.join(data_path, 'subjects'),
+    #     grid_spacing=None)
```
shall we remove it then?

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+
+    # two different ways of saving
+    source_morph_vol.save(os.path.join(tempdir, 'vol'))
+    source_morph_vol.save(os.path.join(tempdir, 'vol.h5'))
```
what are you testing here? is it necessary to write twice?

TommyClausner replied:
the if statement of having a file extension

agramfort commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,275 @@
+
+    # check for invalid file name handling ()
+    with pytest.raises(IOError, match='file .* not found'):
+        read_source_morph(os.path.join(tempdir, '42'))
```
if the error is not raised by our code no need to test it. It will raise a common python error.

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -543,122 +547,145 @@ def test_extract_label_time_course():
+                      smooth=-1,
+                      subjects_dir=subjects_dir)
+
+    with warnings.catch_warnings(record=True):
```
can you add a comment to know what warnings you expect to catch here?

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -543,122 +547,145 @@ def test_extract_label_time_course():
     # will break)
-    pytest.raises(ValueError, stc_to1.morph, subject_from, grade=6,
-                  subjects_dir=subjects_dir)
+    with warnings.catch_warnings(record=True):
```
can you add a comment to know what warnings you expect to catch here?

agramfort commented on mne/tests/test_source_estimate.py:
```diff
@@ -543,122 +547,145 @@ def test_extract_label_time_course():
-                         grade=vertices_to, smooth=12, buffer_size=3,
-                         subjects_dir=subjects_dir)
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter('always')
```
can you add a comment to know what warnings you expect to catch here?

agramfort replied:
also w is not used.

agramfort commented on mne/tests/test_source_space.py:
```diff
@@ -655,6 +656,8 @@ def test_morph_source_spaces():
 
 
+@requires_nibabel()
+@requires_dipy()
```
same remark. You should not force dipy for code that did/does not need it.

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+In MNE-Python, FreeSurfer is used to provide structural information of various
+kinds, for :ref:`source estimation
+<sphx_glr_auto_tutorials_plot_mne_dspm_source_localization.py>`. Thereby a
+subject specific structural MRI will be transformed into various
```
MRI is not transformed to "spherical or inflated brain surfaces".
Maybe
"spherical or inflated brain surfaces are obtained from MRI anatomical images"

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+data between different representations and spaces. See e.g.
+:ref:`sphx_glr_auto_tutorials_plot_morph_stc.py` for information about how to
+use FreeSurfer surface representations to allow functional data to morph
+between different brain maps.
```
brain maps -> subjects?

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+First steps
+===========
+
+After downloading and installing the environment needs to be set up correctly.
```
After downloading and installing,

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+sourcing the setup file:
+
+    >>> export FREESURFER_HOME=/path/to/FreeSurfer
+    >>> source $FREESURFER_HOME/SetUpFreeSurfer.sh
```
>>> is for Python code. Use $ instead to refer to commands in terminal.

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+created upon the reconstruction of the anatomical data. Nevertheless the parent
+directory has to be set beforehand:
+
+    >>> export SUBJECTS_DIR=~/subjects
```
$ not >>>

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+
+    >>> my_subject=sample
+    >>> my_NIfTI=/path/to/NIfTI.nii.gz
+    >>> recon-all -i $my_NIfTI -s $my_subject -all
```
$

agramfort commented on tutorials/plot_background_freesurfer.py:
```diff
@@ -0,0 +1,118 @@
+    >>> recon-all -i $my_NIfTI -s $my_subject -all
+
+, where :math:`i` stands for "input" and :math:`s` for "subject". Executing
+this, will create the folder "~/subjects/sample", were all
```
were -> where

agramfort commented on tutorials/plot_object_sourceestimate.py:
```diff
@@ -0,0 +1,197 @@
+
+
+:class:`mne.SourceEstimate` is the result of a source analysis. MNE-Python
+provides different methods for solving the so called 'inverse problem'. Thereby
```
so-called

agramfort commented on tutorials/plot_object_sourceestimate.py:
```diff
@@ -0,0 +1,197 @@
+Knowing this the :class:`mne.SourceEstimate` (within the MNE ecosystem mostly
+called **stc**) represents the carrier of the new time series data, whereas
+:class:`mne.SourceSpaces` (mostly called **src**) the mapping towards the
+anatomical representation.
```
do you say somewhere that stc is the abbreviation of "source time courses" ?

agramfort commented:
ok I stop here. Go go @TommyClausner ! this is going to be a great feature for the MNE users !

TommyClausner commented:
Hey @agramfort regarding: dipy in the tests:

It's checked when creating SourceMorph + I removed the deprecated functionality for morphing labels + surf... Thus the tests now also use under the hood SourceMorph, requiring dipy... any other way to avoid that?

TommyClausner commented:
Hey thanks @agramfort  for all the comments!! Regarding dipy see above - the others I agree :)

larsoner commented:
> Thus the tests now also use under the hood SourceMorph, requiring dipy... any other way to avoid that?

SourceMorph should only `import dipy` for the volume methods (i.e., nest the imports there). If you are only doing surface smoothing / morphing then you should not need it.

TommyClausner commented:
Ok so everything but surface related functions, correct? (because mixed will use it as well)

larsoner commented:
Yes anything that uses `dipy` volume morphing will require it and that's fine

TommyClausner commented:
Hey I know you don't like **kwargs, but regarding the check of nibabel and dipy... I have to check multiple times in between, if I avoid the general check in the beginning, so I though of a function like the following, to flexibly wrap it:
```
def _check_dep(**kwargs):
    """Wrapper to allow more precise checking instead of one check in
    __init__. Takes pairs of dependency as keyword and version as string and
    returns an ImportError in case the dependency is not satisfied. Example
    _check_dep(nibabel='2.1.0', dipy='0.10.1')
    """
    for lib, ver in kwargs.items():
        passed = check_version(lib, ver)

        if not passed:
            raise ImportError('%s %s or higher must be correctly '
                              'installed and accessible from Python' % lib,
                              ver)
```
one could then do:

```
_check_dep(nibabel='2.1.0')
```
or
```
_check_dep(nibabel='2.1.0', dipy='0.10.1', whateverlibrarymaycometomneaswell='1.0')
```
would this be fine for you?

TommyClausner commented:
Hm... Nevermind I'll do it differently

TommyClausner commented:
Hello CircleCI!

Changes are implemented. Now the shortcuts session is called "additional info" and was slightly rewritten...

The rest was implemented as suggested (I hope) - but I really start getting massively blind on reading the text, so I hope I didn't introduce any (more) language mistakes...

Regarding the warnings: it's a nibabel warning regarding numpy something and a dipy warning regarding numpy version change or so...

I removed most of the "warning catchers" in the tests, because I have them in the class actually...

Here are the renderings (once done):

TommyClausner commented:
Seems the checks fail because I deprecated "mne.morph_data"... Shall I de-deprecate or modify the tests to use SourceMorph ? Because I remember we shortly discussed this ages ago and I'm not quite sure anymore what was the outcome...

So: deprecating ``mne.morph_data`` (is internally wrapped anyway) or keeping it for e.g. convenience?

agramfort commented:
deprecate it and fix the tests.

larsoner commented:
Use pytest.deprecated_call

>

TommyClausner commented:
Why is this conflict in test_source_estimate.py ?
https://github.com/mne-tools/mne-python/pull/5338/conflicts
how can I resolve it?

larsoner commented:
Conflicts arise when your branch changes some lines that have been changed in master since you opened your PR.

To fix you either rebase (preferred) or merge with latest master.

TommyClausner commented:
if I want to rebase it I have to close this PR first I guess(?) - because I pushed from my master(?)

TommyClausner commented:
because this will cause a mess

larsoner commented:
If you plan to rebase, you might want to first squash all of your commits from this branch to a single one

larsoner commented:
You do not need to close this, you can force push instead. But this might not be the best option in case you mess it up

TommyClausner commented:
yeah the problem is that last time I smartly rebased, I commited everything... and now I'm afraid :D

larsoner commented:
(whether you are on master in your fork or not does not make any difference, master is a branch just like any other, except by convention we tend not to mess with it / keep it tracking upstream master)

larsoner commented:
Then yes feel free to close, do your magic, push a new branch and open a new PR

TommyClausner commented:
Aye!

massich commented:
Just for completness, this PR is retaken here https://github.com/mne-tools/mne-python/pull/5430
